src/Json.php

```diff
@@ -70,6 +71,47 @@ public function retrieveValue(?array $data, ?ReflectionNamedType $type = null)
             $data = $data[$pathBit];
         }
 
+        if ($type instanceof ReflectionUnionType) {
```

Comment:
@khepin can you give this another look? couple of new things
 - support null within union type
 - support array OR custom object within union type, not both at the same time because the input structure will look the same. 
 - support a single custom object within union type because the input structure will look the same.